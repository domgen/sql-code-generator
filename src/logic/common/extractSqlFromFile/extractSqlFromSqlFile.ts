import { readFileAsync } from '../readFileAsync';

export const extractSqlFromSqlFile = ({ filePath }: { filePath: string }) => readFileAsync({ filePath }); // the sql is everything in the file in this case
