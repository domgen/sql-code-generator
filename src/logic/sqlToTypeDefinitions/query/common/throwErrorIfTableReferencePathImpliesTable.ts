/**
 * instead of "select id from user" we're asking the user to "select user.id from user" or "select u.id from user u"
 *
 * rationale:
 * - this makes it so that we don't have to try and figure out the implicit reference
 * - this makes it so that when a user does join to another table, they don't have to add the explicit references after the fact (i.e., makes it easier to extend their queries)
 */
export const throwErrorIfTableReferencePathImpliesTable = ({ referencePath }: { referencePath: string }) => {
  if (referencePath.split('.').length !== 2) {
    throw new Error(
      `table reference of "${referencePath}" does not have the table alias explicitly defined, violating best practice`,
    );
  }
};
