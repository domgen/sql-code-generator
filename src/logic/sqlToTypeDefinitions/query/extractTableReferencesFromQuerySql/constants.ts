export const JOIN_TYPES = ['JOIN', 'INNER JOIN', 'LEFT JOIN', 'OUTER JOIN'];
export const TABLE_REFERENCE_TYPE = ['FROM', ...JOIN_TYPES];
