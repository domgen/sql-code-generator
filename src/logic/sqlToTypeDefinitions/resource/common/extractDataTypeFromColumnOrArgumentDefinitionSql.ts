import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (mysqlStringTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.STRING;
  if (mysqlNumberTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.NUMBER;
  if (mysqlDateTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.DATE;
  if (mysqlBinaryTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
