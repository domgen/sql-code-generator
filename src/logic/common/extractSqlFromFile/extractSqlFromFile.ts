import { extractSqlFromSqlFile } from './extractSqlFromSqlFile';
import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const SUPPORTED_FILE_EXTENSIONS = ['.sql', '.ts'];
export const extractSqlFromFile = async ({
  filePath,
}: {
  filePath: string;
}) => {
  // 1. grab extension from file; check that we support it
  const extension = `.${filePath.split('.').slice(-1)[0]}`;
  if (!SUPPORTED_FILE_EXTENSIONS.includes(extension)) {
    throw new Error(`file extension '.${extension}' is not supported`);
  }

  // 2. grab sql from file, based on file type
  if (extension === '.sql') return extractSqlFromSqlFile({ filePath });
  if (extension === '.ts') return extractSqlFromTsFile({ filePath });
  throw new Error('unexpected'); // should never reach here, since we checked earlier that the extension was supported
};
