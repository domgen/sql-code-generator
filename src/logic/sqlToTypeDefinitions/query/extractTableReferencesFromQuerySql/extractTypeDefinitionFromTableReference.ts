import { TypeDefinitionOfQueryTableReference } from '../../../../model/valueObjects/TypeDefinitionOfQueryTableReference';
import { TABLE_REFERENCE_TYPE } from './constants';

const TABLE_NAME_MATCHER_REGEX = `(?:${
  // upper case (e.g., "FROM', 'JOIN', 'LEFT JOIN')
  TABLE_REFERENCE_TYPE.join('|')
}|${
  // and lowercase (e.g., 'from', 'join', 'left join')
  TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join('|')
})\\s(?:\\w+\\.)?(\\w+)?`;

const SPECIFIED_ALIAS_MATCHER_REGEX = `(?:${
  // upper case (e.g., "FROM', 'JOIN', 'LEFT JOIN')
  TABLE_REFERENCE_TYPE.join('|')
}|${
  // and lowercase (e.g., 'from', 'join', 'left join')
  TABLE_REFERENCE_TYPE.map((str) => str.toLowerCase()).join('|')
})\\s(?:\\w+\\.)?(?:\\w+)(?:\\s(?:as|AS))?\\s(\\w+)?`;

export const extractTypeDefinitionFromTableReference = ({ sql }: { sql: string }) => {
  // grab the source path
  const [_, tableName] = new RegExp(TABLE_NAME_MATCHER_REGEX).exec(sql) ?? []; // tslint:disable-line no-unused
  if (!tableName) throw new Error('could not extract the table name from table reference sql; unexpected'); // fail fast

  // grab the alias, if any
  const [__, specifiedAlias] = new RegExp(SPECIFIED_ALIAS_MATCHER_REGEX).exec(sql) ?? []; // tslint:disable-line no-unused

  // define the name, considering whether alias was given
  const alias = specifiedAlias ? specifiedAlias : tableName;

  // return the full definition
  return new TypeDefinitionOfQueryTableReference({
    alias,
    tableName,
  });
};
