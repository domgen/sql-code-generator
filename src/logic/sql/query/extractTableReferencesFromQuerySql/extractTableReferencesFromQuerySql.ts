import { extractTableReferenceSqlSetFromQuerySql } from './extractTableReferenceSqlSetFromQuerySql';
import { extractTypeDefinitionFromTableReference } from './extractTypeDefinitionFromTableReference';

export const extractTableReferencesFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. get the table reference sqls from the query sql
  const tableReferenceSqlSet = extractTableReferenceSqlSetFromQuerySql({ sql });

  // 2. get the type definitions of table references from each table reference sql
  return tableReferenceSqlSet.map((sql) => extractTypeDefinitionFromTableReference({ sql }));
};
