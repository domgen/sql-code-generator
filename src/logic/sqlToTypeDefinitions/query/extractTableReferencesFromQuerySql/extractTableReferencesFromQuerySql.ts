import { extractTypeDefinitionFromTableReference } from './extractTypeDefinitionFromTableReference';
import { tryToExtractTableReferenceSqlSet } from './tryToExtractTableReferenceSqlSetFromQuerySql';
import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';

export const extractTableReferencesFromQuerySql = ({ sql }: { sql: string }) => {
  // 0. tokenize each subquery
  const { flattenedSql } = flattenSqlByReferencingAndTokenizingSubqueries({ sql }); // flatten the sql - so that subqueries don't hinder regular parsing

  // 1. get the table reference sqls from the query sql
  const tableReferenceSqlSet = tryToExtractTableReferenceSqlSet({ sql: flattenedSql }); // "try to", since its okay if there are no table references; (e.g., `select upsert_something(...)`)

  // 2. get the type definitions of table references from each table reference sql
  return tableReferenceSqlSet.map((sql) => extractTypeDefinitionFromTableReference({ sql }));
};
