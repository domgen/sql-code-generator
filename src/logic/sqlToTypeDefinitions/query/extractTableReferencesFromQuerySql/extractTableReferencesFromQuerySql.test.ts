import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTableReferencesFromQuerySql } from './extractTableReferencesFromQuerySql';
import { TypeDefinitionOfQueryTableReference } from '../../../../model';

describe('extractTableReferencesFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/select_suggestion.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for an example joining to multiple tables', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_users_by_last_name.sql`,
    });
    const defs = extractTableReferencesFromQuerySql({ sql });
    expect(defs.length).toEqual(3);
    expect(defs).toContainEqual(
      new TypeDefinitionOfQueryTableReference({ alias: 'up', tableName: 'view_user_profile_current' }),
    );
    expect(defs).toMatchSnapshot();
  });
});
