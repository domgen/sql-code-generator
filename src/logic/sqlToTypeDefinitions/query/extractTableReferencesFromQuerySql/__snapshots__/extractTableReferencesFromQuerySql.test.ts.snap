// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for this example 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "i",
    "tableName": "image",
  },
]
`;

exports[`extractTableReferencesFromQuerySql should be able to determine types accurately for this other example 1`] = `
Array [
  TypeDefinitionOfQueryTableReference {
    "alias": "s",
    "tableName": "suggestion",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "c",
    "tableName": "suggestion_cvp",
  },
  TypeDefinitionOfQueryTableReference {
    "alias": "v",
    "tableName": "suggestion_version",
  },
]
`;
